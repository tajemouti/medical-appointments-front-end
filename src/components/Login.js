import { Button, CircularProgress, TextField } from '@mui/material';
import { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link, useNavigate } from 'react-router-dom';
import img from '../images/bg.svg';
import { loginUser } from '../features/user/userSlice';

function Login() {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const loginResponse = useSelector((state) => state.user.user.token);
  const [loading, setLoading] = useState(false);

  const [userName, setUserName] = useState('');
  const [password, setPassword] = useState('');

  const handleUsernameChange = (event) => {
    setUserName(event.target.value);
  };

  const handlePasswordChange = (event) => {
    setPassword(event.target.value);
  };

  useEffect(() => {
    if (loginResponse) {
      navigate('/');
    }
  }, [loginResponse, navigate]);

  const handleSubmit = async (event) => {
    event.preventDefault();
    setLoading(true);

    try {
      await dispatch(loginUser({
        username: userName,
        password,
      }));
    } finally {
      setLoading(false);
    }
  };

  return (
    <>
      <div className="flex items-center justify-center h-screen bg-cover bg-center" style={{ backgroundImage: `url(${img})` }}>
        <div className="flex items-center justify-center flex-col gap-6 w-4/5 p-12 md:max-w-fit md:max-h-fit bg-white rounded-md">
          <h1 className="text-gray-800 font-bold text-2xl">
            Welcome to
            <span className="text-[#97BF0F]"> Health Clinic</span>
          </h1>
          <h2 className="text-gray-800 font-bold text-2xl">Log in</h2>
          <form className="flex items-center justify-center flex-col gap-6" onSubmit={(e) => handleSubmit(e)}>
            <TextField
              required
              id="outlined-basics"
              type="text"
              value={userName}
              onChange={(e) => handleUsernameChange(e)}
              label="Username"
              variant="outlined"
            />
            <TextField
              required
              id="outlined-basic"
              type="password"
              value={password}
              onChange={(e) => handlePasswordChange(e)}
              label="Password"
              variant="outlined"
            />
            <div className="flex gap-4">
              <Button type="submit" variant="outlined" disabled={loading}>
                {loading ? <CircularProgress size={24} /> : 'Login'}
              </Button>
              <Button variant="outlined">
                <Link to="/signup">Sign up</Link>
              </Button>
            </div>
          </form>
        </div>
      </div>
    </>
  );
}

export default Login;
