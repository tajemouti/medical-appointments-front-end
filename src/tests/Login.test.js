import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Login from '../components/Login';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => jest.fn(),
}));

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate,
}));

let num = 0;
const handleClick = (n) => {
  num = n;
};
describe('Tests login component', () => {
  it('should render page', () => {
    const loginComponent = renderer.create(
      <BrowserRouter>
        <Login />
      </BrowserRouter>,
    );
    expect(loginComponent).toMatchSnapshot();
  });

  it('should isplay "Welcome to Health Clinic"', () => {
    render(
      <BrowserRouter>
        <Login />
      </BrowserRouter>,
    );

    const h1 = screen.getByText('Welcome to');
    const h2 = screen.getByText('Health Clinic');
    expect(h1).toBeInTheDocument();
    expect(h2).toBeInTheDocument();
  });

  it('should allow click on signup button', () => {
    render(
      <BrowserRouter>
        <Login />
      </BrowserRouter>,
    );

    const loginButton = screen.getByText('Login');
    loginButton.onclick = handleClick(2);
    fireEvent.click(loginButton);
    expect(num).toBe(2);
  });

  it('should allow click on signup button', () => {
    render(
      <BrowserRouter>
        <Login />
      </BrowserRouter>,
    );

    const signUpButton = screen.getByText('Sign up');
    signUpButton.onclick = handleClick(1);
    fireEvent.click(signUpButton);
    expect(num).toBe(1);
  });

  it('should get value from username field', () => {
    render(
      <BrowserRouter>
        <Login />
      </BrowserRouter>,
    );

    const usernameInput = screen.getByRole('textbox');
    expect(usernameInput).toBeInTheDocument();
    fireEvent.change(usernameInput, { target: { value: 'testuser' } });
    expect(usernameInput.value).toBe('testuser');
  });
});
