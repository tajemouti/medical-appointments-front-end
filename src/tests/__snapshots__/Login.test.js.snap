// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests login component should render page 1`] = `
<div
  className="flex items-center justify-center h-screen bg-cover bg-center"
  style={
    Object {
      "backgroundImage": "url(bg.svg)",
    }
  }
>
  <div
    className="flex items-center justify-center flex-col gap-6 w-4/5 p-12 md:max-w-fit md:max-h-fit bg-white rounded-md"
  >
    <h1
      className="text-gray-800 font-bold text-2xl"
    >
      Welcome to
      <span
        className="text-[#97BF0F]"
      >
         Health Clinic
      </span>
    </h1>
    <h2
      className="text-gray-800 font-bold text-2xl"
    >
      Log in
    </h2>
    <form
      className="flex items-center justify-center flex-col gap-6"
      onSubmit={[Function]}
    >
      <div
        className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="outlined-basics"
          id="outlined-basics-label"
        >
          Username
          <span
            aria-hidden={true}
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id="outlined-basics"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-yjsfm1"
            >
              <span>
                Username
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="outlined-basic"
          id="outlined-basic-label"
        >
          Password
          <span
            aria-hidden={true}
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id="outlined-basic"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="password"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-yjsfm1"
            >
              <span>
                Password
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        className="flex gap-4"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="submit"
        >
          Login
        </button>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <a
            href="/signup"
            onClick={[Function]}
          >
            Sign up
          </a>
        </button>
      </div>
    </form>
  </div>
</div>
`;
